### PR TITLE
feat: Add C35 a supervision order (FPLA-2922)

### PR DESCRIPTION
**Pull request by jacmurphy**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-2922

### Change description ###

- [x] Add option in menu
- [x] Add new Supervision order details page
- [x] Add fields to store information provided
- [x] Create a validator for data
- [x] Create a docmosis template 
- [x] Address docmosis updates from ticket 
- [x] Add E2E test

### Acceptance Criteria ###
Description

As an Admin or judge
I want to create supervision orders
So that I can resolve the case quickly at final order

Acceptance Criteria

    Add new option to list of orders on select order page:
        Supervision order (C35A)
        Ensure the list is in Alphabetical order with the order name and the number in brackets afterwards)
    On the approval/issuing details page, pre-populate the approval date with current date and allow user to edit (this should behave the same for all order types)
    Other order details for supervision order:
        Date options
            Number of months from date of order
            Until a named date
            Until a specific time on a named date
        Validate that the end date is no more than 12 months from order start (Supervision orders cannot last for more than 12 months)
        Add further directions? (Optional)
    On completion of the event:
        Store pre-sealed Word version of the order (for slip rule amendments later)
        Seal the order
        Convert to pdf
        Display order in the Orders tab (most recent order at the top)
    Tab display details:
        Order type (Supervision order)
        Order approval date
        Order duration or end date ?
    Send order issued notifications to parties according to serving preference
    Toggle off for release

Order flow
Create an order > Select Supervision order > Issuing details (judge and approval date) > Select children > Other details page (enter date/time/duration); enter additional directions if required) > Preview order > CYA

Order output
The Court orders [Name of Local Authority] to supervise the child/[children]
#for [NUMBER_OF_MONTHS] months from the date of this order until [CALCULATED_DATE].
#until [ON_A_SET_DAY].
#until [ON_A_SET_TIME] on [ON_A_SET_DAY].

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/IssuingDetailsApprovalDateBlockPrePopulator.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import uk.gov.hmcts.reform.fpl.service.time.Time;
+
+import java.time.LocalDate;
+import java.util.Map;
+
+import static uk.gov.hmcts.reform.fpl.model.order.OrderQuestionBlock.APPROVAL_DATE;
+
+@Component
+@RequiredArgsConstructor(onConstructor_ = {@Autowired})
+public class IssuingDetailsApprovalDateBlockPrePopulator implements QuestionBlockOrderPrePopulator {
```
Comment (samuelzerozero) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/IssuingDetailsApprovalDateBlockPrePopulator.java:17]: this prepopulator when Approval date is there, is that the intention?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidatorTest.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.List;
+
+import static java.util.Objects.deepEquals;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static uk.gov.hmcts.reform.fpl.enums.orders.SupervisionOrderEndDateType.SET_CALENDAR_DAY;
+import static uk.gov.hmcts.reform.fpl.enums.orders.SupervisionOrderEndDateType.SET_CALENDAR_DAY_AND_TIME;
+import static uk.gov.hmcts.reform.fpl.enums.orders.SupervisionOrderEndDateType.SET_NUMBER_OF_MONTHS;
+import static uk.gov.hmcts.reform.fpl.model.order.OrderQuestionBlock.SUPERVISION_ORDER_END_DATE;
+
+class SupervisionOrderEndDateValidatorTest {
```
Comment (samuelzerozero) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidatorTest.java:21]: missing the positive tests and would be good to have boundary tests. For example if you know the precision is at second, you   .manageOrdersSetDateEndDate(dateNow().plusDays(1))) for the negative and  .manageOrdersSetDateEndDate(dateNow())) for the positive (or something along the line....)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/IssuingDetailsApprovalDateBlockPrePopulator.java (hunk 1 above)
Comment (jacmurphy) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/IssuingDetailsApprovalDateBlockPrePopulator.java:17]: Yes, I have an AC that asked for this to be added 
```On the approval/issuing details page, pre-populate the approval date with current date and allow user to edit (this should behave the same for all order types)```

File: ccd-definition/AuthorisationCaseField/CareSupervision/court-admin.json [hunk 3]
```diff
@@ -2085,6 +2085,34 @@
     "UserRole": "caseworker-publiclaw-courtadmin",
     "CRUD": "R"
   },
+  {
```
Comment (samuelzerozero) [inline ccd-definition/AuthorisationCaseField/CareSupervision/court-admin.json:2088]: this need to be replicated for the other users. would follow what's been done for the other users (i-e judges...)

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidatorTest.java (hunk 2 above)
Comment (jacmurphy) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidatorTest.java:21]: Excellent point. Ill get into it now

File: ccd-definition/AuthorisationCaseField/CareSupervision/court-admin.json (hunk 3 above)
Comment (jacmurphy) [reply ccd-definition/AuthorisationCaseField/CareSupervision/court-admin.json:2088]: Ah yes, you're absolutely right. Adding now

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidatorTest.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        CaseData caseData = CaseData.builder()
+            .manageOrdersEventData(ManageOrdersEventData.builder()
+                .manageSupervisionOrderEndDateType(SET_CALENDAR_DAY_AND_TIME)
+                .manageOrdersSetDateAndTimeEndDate(invalidTime)
+                .build())
+            .build();
+        assertThat(underTest.validate(caseData)).isEqualTo(List.of(TEST_INVALID_TIME_MESSAGE));
+    }
+
+    private LocalDate dateNow() {
```
Comment (samuelzerozero) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidatorTest.java:124]: this can be a constant NOW (since it's fixed time)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGenerator.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.List;
+
+import static uk.gov.hmcts.reform.fpl.utils.DateFormatterHelper.DATE_TIME_WITH_ORDINAL_SUFFIX;
+import static uk.gov.hmcts.reform.fpl.utils.DateFormatterHelper.DATE_WITH_ORDINAL_SUFFIX;
+import static uk.gov.hmcts.reform.fpl.utils.DateFormatterHelper.formatLocalDateTimeBaseUsingFormat;
+import static uk.gov.hmcts.reform.fpl.utils.DateFormatterHelper.getDayOfMonthSuffix;
+
+@Component
+@RequiredArgsConstructor(onConstructor_ = {@Autowired})
+public class C35aSupervisionOrderDocumentParameterGenerator implements DocmosisParameterGenerator {
```
Comment (samuelzerozero) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGenerator.java:30]: Missing unit test for this one

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/event/ManageOrdersEventData.java [hunk 6]
```diff
@@ -30,4 +31,8 @@
     DocumentReference manageOrdersPowerOfArrest;
     Address manageOrdersEpoRemovalAddress;
     LocalDate manageOrdersExclusionStartDate;
+    SupervisionOrderEndDateType manageSupervisionOrderEndDateType;
+    LocalDate manageOrdersSetDateEndDate;
+    LocalDateTime manageOrdersSetDateAndTimeEndDate;
+    int manageOrdersSetMonthsEndDate;
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/event/ManageOrdersEventData.java:37]: Updated

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGenerator.java (hunk 5 above)
Comment (jacmurphy) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGenerator.java:31]: File added

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java [hunk 7]
```diff
@@ -0,0 +1,282 @@
+package uk.gov.hmcts.reform.fpl.service.orders.generator;
```
Comment (jacmurphy) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java:1]: @samuelzerozero This is the test file you requested

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/IssuingDetailsApprovalDateBlockPrePopulator.java (hunk 1 above)
Comment (samuelzerozero) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/IssuingDetailsApprovalDateBlockPrePopulator.java:17]: ah sorry, got confused by the name. Probably ApprovalDateBlockPrePopulator?

**Review by samuelzerozero — APPROVED** [review]

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidator.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+@Component
+@RequiredArgsConstructor(onConstructor_ = {@Autowired})
+public class SupervisionOrderEndDateValidator implements QuestionBlockOrderValidator {
+    private static final Integer MINIMUM_MONTHS_VALID = 1;
+    private static final Integer MAXIMUM_MONTHS_VALID = 12;
+    private static final String INVALID_TIME_MESSAGE = "Enter a valid time";
+    private static final String FUTURE_DATE_MESSAGE = "Enter an end date in the future";
+    private static final String END_DATE_MAX_RANGE_MESSAGE = "Supervision orders cannot last longer than 12 months";
+    private static final String END_DATE_MIN_RANGE_MESSAGE = "*******Supervision orders in months should be at least 1";
```
Comment (Rob-Doherty) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidator.java:28]: did you mean to include the `*******` at the start of this message?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+            NEXT_WEEK_DATE_TIME,
+            String.format(DATE_WITH_ORDINAL_SUFFIX, dayOrdinalSuffix)
+        );
+        String courtOrderMessage = "The Court orders " + LA_NAME
+            + " supervises the " + CHILDREN_GRAMMAR
+            + " until " + formattedDate + ".";
+
+        when(laNameLookup.getLocalAuthorityName(LA_CODE)).thenReturn(LA_NAME);
+
+        List<Element<Child>> selectedChildren = wrapElements(CHILD, CHILD);
```
Comment (Rob-Doherty) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java:84]: this test title is for a single child, but it looks like you've created a scenario with multiple children. should this be the following instead?
```suggestion
        List<Element<Child>> selectedChildren = wrapElements(CHILD);
```

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+
+        when(childrenService.getSelectedChildren(caseData)).thenReturn(selectedChildren);
+
+        DocmosisParameters generatedParameters = underTest.generate(caseData);
+        DocmosisParameters expectedParameters = expectedCommonParameters()
+            .orderDetails(courtOrderMessage)
+            .build();
```
Comment (Rob-Doherty) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java:91]: are the expected details verifying the correct children are listed? so when it's a single child case that only 1 child is present in the document? or is that covered somewhere else?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidatorTest.java [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    @Test
+    void shouldReturnErrorForNumberOfMonthsIsLessThanAccepted() {
+        CaseData caseData = CaseData.builder()
+            .manageOrdersEventData(ManageOrdersEventData.builder()
+                .manageSupervisionOrderEndDateType(SET_NUMBER_OF_MONTHS)
+                .manageOrdersSetMonthsEndDate(MINIMUM_MONTHS_ACCEPTED - 1)
+                .build())
+            .build();
+
+        deepEquals(underTest.validate(caseData), TEST_UNDER_DATE_RANGE_MESSAGE);
```
Comment (Rob-Doherty) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidatorTest.java:191]: is `deepEquals()` working the way you expect here? it didn't pick up the different error message in `SupervisionOrderEndDateValidator` so I wonder if it's giving false posities

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidator.java (hunk 8 above)
Comment (jacmurphy) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidator.java:28]: Yes, it seems to be a styling markup in CCD

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java [hunk 12]
```diff
@@ -314,6 +333,60 @@ void shouldThrowExceptionWhenMidEventUrlParameterDoesNotMatchSectionNames() {
...
+            .manageOrdersEventData(ManageOrdersEventData.builder()
+                .manageOrdersType(C35A_SUPERVISION_ORDER)
+                .manageSupervisionOrderEndDateType(SET_NUMBER_OF_MONTHS)
+                .manageOrdersSetMonthsEndDate(testInvalidMonth)
+                .build())
+            .build();
+
+        AboutToStartOrSubmitCallbackResponse response = postMidEvent(caseData, "order-details");
+
+        assertThat(response.getErrors()).containsOnly(testEndDateRangeMessage);
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java:369]: Firstly, thank you @Rob-Doherty  for looking into the code while testing. Its a lovely thing to see!

There are 3 types of ways to set the date:
DATE, DATETIME, Integer

In this scenario, i was testing the integer, which is related directly to Month. So the testCase of incrimenting the integer would not lend itself to the possiblity of adding a day to the end of it

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java (hunk 9 above)
Comment (jacmurphy) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java:84]: You're 100% correct, well spotted. Ill update it in the next commit that should retrigger its build

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidatorTest.java [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+                .manageOrdersSetDateAndTimeEndDate(onBoundaryDateTime)
+                .build())
+            .build();
+
+        assertThat(underTest.validate(caseData)).isEqualTo(List.of());
+    }
+
+    @Test
+    void shouldRejectDateTimeOverBoundaryOfMaxAllowedEndDateTime() {
+        LocalDateTime overBoundaryDateTime = time.now().plusYears(1).plusSeconds(1);
```
Comment (Rob-Doherty) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidatorTest.java:107]: great boundary testing 👍

File: e2e/tests/adminManagesOrders_test.js [hunk 14]
```diff
@@ -133,6 +132,32 @@ Scenario('Create C21 blank order', async ({I, caseViewPage, manageOrdersEventPag
...
+  await I.goToNextPage();
+  await manageOrdersEventPage.selectOrder(manageOrdersEventPage.orders.options.c35a);
+  await I.goToNextPage();
+  await manageOrdersEventPage.enterJudge();
+  await I.goToNextPage();
+  await manageOrdersEventPage.selectChildren(manageOrdersEventPage.section3.allChildren.options.select, [0]);
+  await I.goToNextPage();
+  await manageOrdersEventPage.enterFurtherDirections('Supervision order further details.');
+  await manageOrdersEventPage.selectSupervisionOrder(manageOrdersEventPage.section4.supervisionOrderType.options.numberOfMonths);
+  await manageOrdersEventPage.enterSuperVisionNumOfMonths(10);
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply e2e/tests/adminManagesOrders_test.js:147]: Great sugestion. Updating in next commit

File: ccd-definition/FixedLists/CareSupervision/ManageOrdersType.json [hunk 15]
```diff
@@ -3,21 +3,28 @@
...
     "ListElementCode": "C23_EMERGENCY_PROTECTION_ORDER",
-    "ListElement": "C23 - Emergency protection order",
+    "ListElement": "Emergency protection order (C23)",
     "DisplayOrder": 3
+  },
+  {
+    "LiveFrom": "01/01/207",
+    "ID": "ManageOrdersType",
+    "ListElementCode": "C35A_SUPERVISION_ORDER",
+    "ListElement": "Supervision order (C35a)",
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply ccd-definition/FixedLists/CareSupervision/ManageOrdersType.json:27]: Youre correct. Updating in next commit

File: e2e/tests/adminManagesOrders_test.js [hunk 16]
```diff
@@ -133,6 +132,32 @@ Scenario('Create C21 blank order', async ({I, caseViewPage, manageOrdersEventPag
...
+  await I.goToNextPage();
+  await manageOrdersEventPage.checkPreview();
+  await I.completeEvent('Save and continue');
+  I.seeEventSubmissionConfirmation(config.administrationActions.manageOrders);
+  assertOrder(I, caseViewPage, {
+    orderIndex: 3,
+    orderType: 'C35a - Supervision order',
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply e2e/tests/adminManagesOrders_test.js]: Updating in next commit

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jacmurphy) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java]: I have done a rework on all these tests and they are a lot more comprehensive now

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java [hunk 17]
```diff
@@ -314,6 +333,60 @@ void shouldThrowExceptionWhenMidEventUrlParameterDoesNotMatchSectionNames() {
...
+                .build())
+            .build();
+
+        AboutToStartOrSubmitCallbackResponse response = postMidEvent(caseData, "order-details");
+
+        assertThat(response.getErrors()).containsOnly(testFutureDateMessage);
+    }
+
+    @Test
+    void supervisionOrderNumberOfMonthsShouldNotAllowInvalidFutureDate() {
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java:356]: I have done a rework on all these tests and they are a lot more comprehensive now

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidator.java (hunk 8 above)
Comment (jacmurphy) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/SupervisionOrderEndDateValidator.java]: this was initially put it as i beleived it was a formatting necessity but now that i know it is not, its been removed

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java (hunk 9 above)
Comment (jacmurphy) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java]: updated

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java (hunk 10 above)
Comment (jacmurphy) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java:93]: There currently arent tests to show the correct children are listed. The logic used to display the children is the same used elsewhere in the application and my primary focus was the content of when the order ends. Do you see these test cases as a necessity ?

Comment (Rob-Doherty) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C35aSupervisionOrderDocumentParameterGeneratorTest.java:93]: no that's fair enough, if we've got the coverage at a lower level, and this test was just about proving the API comms then that makes sense

**Review by nraghu32 — APPROVED** [review]
